test/runtests.jl

```diff
@@ -289,120 +283,183 @@ end
-        obs[i] = height[stations.ist[i], stations.jst[i],1] + rand()
+    seed = 1689017430981346891
+    MPI.Init()
+    my_rank = 0
+    input_file = joinpath(
+        dirname(pathof(ParticleDA)), "..", "test", "optimal_filter_test_2.yaml"
+    )
+    filter_type = OptimalFilter()
+    rng = Random.seed!(seed)
+    user_input_dict = ParticleDA.read_input_file(input_file)
+    model_params_dict = get(user_input_dict, "model", Dict())
+    filter_params_dict = get(user_input_dict, "filter", Dict())
+    # Compute state noise covariance matrix once only as potentially large and invariant
+    # to number of particles
+    model_data = Model.init(model_params_dict, 1, my_rank, rng)
+    grid_size = ParticleDA.get_grid_size(model_data)
+    cell_size = ParticleDA.get_grid_cell_size(model_data)
+    indices = ParticleDA.get_observed_state_indices(model_data)
+    state_noise_covariance_structure = ParticleDA.get_model_noise_params(model_data)
+    grid_coord_1 = repeat((1:grid_size[1]) .* cell_size[1], outer=grid_size[2])
+    grid_coord_2 = repeat((1:grid_size[2]) .* cell_size[2], inner=grid_size[1])
+    cov_X_X = ParticleDA.state_noise_covariance.(
+        abs.(grid_coord_1 .- grid_coord_1'),
+        abs.(grid_coord_2 .- grid_coord_2'),
+        (state_noise_covariance_structure,),
+    )
+    for nprt in [25, 100, 400, 2500]
+        filter_params_dict["nprt"] = nprt
+        filter_params = ParticleDA.get_params(FilterParameters, filter_params_dict)
+        nprt_per_rank = nprt
+        model_data = Model.init(model_params_dict, nprt_per_rank, my_rank, rng)
+        filter_data = ParticleDA.init_filter(
+            filter_params, model_data, nprt_per_rank, rng, Float64, filter_type
+        )
+        observations = ParticleDA.update_truth!(model_data, nprt_per_rank)
+        initial_particles = copy(ParticleDA.get_particles(model_data))
+        firstonlastaxis(array) = selectdim(array, ndims(array), 1)
+        # force all particles equal to first
+        initial_particles .= firstonlastaxis(initial_particles)
+        ParticleDA.set_particles!(model_data, initial_particles)
+        ParticleDA.update_particle_dynamics!(model_data, nprt_per_rank)
+        particles = copy(ParticleDA.get_particles(model_data))
+        # update_particle_dynamics should at least some change particle components
+        @test any(particles .!= initial_particles)
+        # update_particle_dynamics should be deterministic so all particles remain equal 
+        # to each other
+        @test all(firstonlastaxis(particles) .== particles)
+        observations_mean_given_particle = copy(firstonlastaxis(
+            ParticleDA.get_particle_observations!(model_data, nprt_per_rank)
+        ))
+        ParticleDA.update_particle_noise!(model_data, nprt_per_rank)
+        noised_particles = copy(ParticleDA.get_particles(model_data))
+        # update_particle_noise! should add noise to all particle components
+        @test all(particles .!= noised_particles)
```

Comment:
I guess that technically because of limited floating point-precision there is a non-zero probability of one or more noise components being exactly zero and for a large enough number of dimensions it could become non-negligible, though I think it would need to be very high: even for `Float16` precision it seems to take ~10⁷ draws from `randn` on average to get an exactly zero value (and my intuition would be for `Float64` it will take roughly 2⁴⁸ times as many draws on average).

Given the state noise covariance is being computed explicitly I can make this test more specific to check that the noise is approximately zero mean and of correct covariance to within Monte Carlo error.

Reply:
Now removed in favour of checking state noise (of observed state components) has correct covariance.